internal/app/machined/pkg/runtime/drainer.go

```diff
@@ -0,0 +1,99 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package runtime
+
+import (
+	"context"
+	"sync"
+)
+
+// NewDrainer creates new drainer.
+func NewDrainer() *Drainer {
+	return &Drainer{
+		shutdown: make(chan struct{}, 1),
+	}
+}
+
+// Drainer is used in controllers to ensure graceful shutdown.
+type Drainer struct {
+	subscriptionsMu sync.Mutex
+	subscriptions   []*DrainSubscription
+	shutdown        chan struct{}
+}
+
+// Drain initializes drain sequence waits for it to succeed until the context is canceled.
+func (d *Drainer) Drain(ctx context.Context) error {
+	d.subscriptionsMu.Lock()
+	for _, s := range d.subscriptions {
+		select {
+		case s.events <- DrainEvent{}:
+		default:
+		}
+	}
+	d.subscriptionsMu.Unlock()
+
+	for {
+		select {
+		case <-d.shutdown:
+			d.subscriptionsMu.Lock()
+			l := len(d.subscriptions)
+			d.subscriptionsMu.Unlock()
+
+			if l == 0 {
+				return nil
+			}
+		case <-ctx.Done():
+			return context.Canceled
```

Comment:
this and some places around: `return ctx.Err()` sounds better, as this will return a specific error with the context, e.g. it can be DeadlineExceeded